Detectors/MUON/MCH/Workflow/src/DataDecoder.cxx

```diff
@@ -0,0 +1,602 @@
+                      deId, padId, digitadc, time, stopTime);
+    ++ndigits;
+  };
+
+  const auto patchPage = [&](gsl::span<const std::byte> rdhBuffer) {
+    auto rdhAnyPtr = reinterpret_cast<o2::header::RDHAny*>(const_cast<std::byte*>(&rdhBuffer[0]));
+    if (mRdhHandler) {
+      mRdhHandler(rdhAnyPtr);
+    }
+    auto rdhPtr = reinterpret_cast<o2::header::RAWDataHeaderV4*>(const_cast<std::byte*>(&rdhBuffer[0]));
+    auto& rdh = *rdhPtr;
+    mNrdhs++;
+
+    auto cruId = o2::raw::RDHUtils::getCRUID(*rdhAnyPtr);
+    auto endPointID = o2::raw::RDHUtils::getEndPointID(*rdhAnyPtr);
+    o2::raw::RDHUtils::setFEEID(*rdhAnyPtr, cruId * 2 + endPointID);
+    isStopRDH = o2::raw::RDHUtils::getStop(*rdhAnyPtr);
+
+    feeId = o2::raw::RDHUtils::getFEEID(*rdhAnyPtr);
+    orbit = o2::raw::RDHUtils::getHeartBeatOrbit(*rdhAnyPtr);
+    linkId = o2::raw::RDHUtils::getLinkID(*rdhAnyPtr);
+
+    if (!mMerger) {
+      if (linkId == 15) {
+        mMerger = new Merger;
+      } else {
+        mMerger = new MergerSimple;
+      }
+
+      mMerger->setDigitHandler(storeDigit);
+    }
+    mMerger->setOrbit(feeId, orbit, isStopRDH);
+  };
+
+  if (!mDecoder) {
+    mDecoder = mFee2Solar ? o2::mch::raw::createPageDecoder(page, channelHandler, mFee2Solar)
+                          : o2::mch::raw::createPageDecoder(page, channelHandler);
+  }
+
+  patchPage(page);
+
+  mDecoder(page);
+
+  mMerger->mergeDigits(feeId);
+
+  if (mPrint) {
+    std::cout << "[decodeBuffer] mOutputDigits size: " << mOutputDigits.size() << std::endl;
+    for (auto d : mOutputDigits) {
+      if (d.getPadID() < 0)
+        continue;
+      const Segmentation& segment = segmentation(d.getDetID());
+      float X = segment.padPositionX(d.getPadID());
+      float Y = segment.padPositionY(d.getPadID());
+      bool bend = !segment.isBendingPad(d.getPadID());
+      if (bend)
+        continue;
+      std::cout << fmt::format("  DE {:4d}  PAD {:5d}  ADC {:6d}  TIME ({} {} {:4d})",
+                               d.getDetID(), d.getPadID(), d.getADC(), d.getTime().orbit, d.getTime().bunchCrossing, d.getTime().sampaTime);
+      std::cout << fmt::format("\tC {}  PAD_XY {:+2.2f} , {:+2.2f}", (int)bend, X, Y);
+      std::cout << std::endl;
+    }
+    for (auto d : mOutputDigits) {
+      if (d.getPadID() < 0)
+        continue;
+      const Segmentation& segment = segmentation(d.getDetID());
+      float X = segment.padPositionX(d.getPadID());
+      float Y = segment.padPositionY(d.getPadID());
+      bool bend = !segment.isBendingPad(d.getPadID());
+      if (!bend)
+        continue;
+      std::cout << fmt::format("  DE {:4d}  PAD {:5d}  ADC {:6d}  TIME ({} {} {:4d})",
+                               d.getDetID(), d.getPadID(), d.getADC(), d.getTime().orbit, d.getTime().bunchCrossing, d.getTime().sampaTime);
+      std::cout << fmt::format("\tC {}  PAD_XY {:+2.2f} , {:+2.2f}", (int)bend, X, Y);
+      std::cout << std::endl;
+    }
+    for (auto d : mOutputDigits) {
```

Comment:
in principle this final loop should be empty, right ? No digit should have an invalid padId ?

Reply:
Definitely. I must have introduced it for debugging purposes. I removed it.